File: hash/java/src/main/java/io/github/wmetcalf/rosettasquint/hash/internal/PilGaussianBlur.java
package io.github.wmetcalf.rosettasquint.hash.internal;

/**
 * Pillow-compatible GaussianBlur approximation via 3 separable box filters.
 *
 * Matches PIL.ImageFilter.GaussianBlur(radius=2) exactly:
 *   1. Compute float box radius using _gaussian_blur_radius(2.0, 3) from BoxBlur.c.
 *   2. Apply 3 horizontal passes (ImagingLineBoxBlur8) over rows.
 *   3. Transpose, apply 3 horizontal passes, transpose back (vertical passes).
 *   4. Values are integer-rounded inside each 1-D line via fixed-point (ww/fw).
 *
 * Reference: src/libImaging/BoxBlur.c in Pillow 10.4.0.
 */
public final class PilGaussianBlur {

    private static final int PASSES = 3;
    /** Default PIL GaussianBlur radius (sigma). */
    private static final double RADIUS = 2.0;

    /**
     * Apply PIL GaussianBlur(radius=2) to a uint8 grayscale image.
     *
     * @param src  row-major uint8 array [H][W], values 0-255
     * @return new [H][W] array with blur applied
     */
    public static int[][] apply(int[][] src) {
        int H = src.length;
        int W = src[0].length;

        double floatRadius = gaussianBlurRadius(RADIUS, PASSES);
        int radius = (int) floatRadius;  // truncate toward zero
        long ww = (long) ((1L << 24) / (floatRadius * 2.0 + 1.0));
        long fw = ((1L << 24) - (long) (radius * 2 + 1) * ww) / 2L;

        // --- 3 horizontal passes ---
        int[][] buf = copyOf(src, H, W);
        for (int pass = 0; pass < PASSES; pass++) {
            int[][] next = new int[H][W];
            int edgeA = Math.min(radius + 1, W);
            int edgeB = Math.max(W - radius - 1, 0);
            for (int y = 0; y < H; y++) {
                blurLine8(buf[y], next[y], W - 1, radius, edgeA, edgeB, ww, fw);
            }
            buf = next;
        }

        // --- Transpose ---
        int[][] t = transpose(buf, H, W);

        // --- 3 horizontal passes on transposed (= vertical) ---
        int edgeAv = Math.min(radius + 1, H);
        int edgeBv = Math.max(H - radius - 1, 0);
        for (int pass = 0; pass < PASSES; pass++) {
            int[][] next = new int[W][H];
            for (int y = 0; y < W; y++) {
                blurLine8(t[y], next[y], H - 1, radius, edgeAv, edgeBv, ww, fw);
            }
            t = next;
        }

        // --- Transpose back ---
        return transpose(t, W, H);
    }

    /**
     * From BoxBlur.c: _gaussian_blur_radius(radius, passes).
     * Computes the float box radius that approximates a Gaussian with std-dev = radius.
     */
    static double gaussianBlurRadius(double radius, int passes) {
        double sigma2 = radius * radius / passes;
        double L = Math.sqrt(12.0 * sigma2 + 1.0);
        double l = Math.floor((L - 1.0) / 2.0);
        double a = (2.0 * l + 1.0) * (l * (l + 1.0) - 3.0 * sigma2);
        a /= 6.0 * (sigma2 - (l + 1.0) * (l + 1.0));
        return l + a;
    }

    /**
     * ImagingLineBoxBlur8 — exact translation of the C function for 'L' (uint8) images.
     *
     * @param lineIn  source row (length >= lastx+1)
     * @param lineOut destination row
     * @param lastx   last valid index (= width - 1)
     * @param radius  integer part of box radius
     * @param edgeA   min(radius+1, width)
     * @param edgeB   max(width-radius-1, 0)
     * @param ww      fixed-point weight for interior pixels
     * @param fw      fixed-point weight for fractional far pixels
     */
    static void blurLine8(int[] lineIn, int[] lineOut,
                           int lastx, int radius, int edgeA, int edgeB,
                           long ww, long fw) {
        // Init acc for position -1 (one to the left of pixel 0)
        long acc = (long) lineIn[0] * (radius + 1);
        for (int x = 0; x < edgeA - 1; x++) {
            acc += lineIn[x];
        }
        acc += (long) lineIn[lastx] * (radius - edgeA + 1);

        if (edgeA <= edgeB) {
            for (int x = 0; x < edgeA; x++) {
                acc += lineIn[x + radius] - lineIn[0];
                long bulk = acc * ww + (long) (lineIn[0] + lineIn[x + radius + 1]) * fw;
                lineOut[x] = (int) ((bulk + (1L << 23)) >> 24);
            }
            for (int x = edgeA; x < edgeB; x++) {
                acc += lineIn[x + radius] - lineIn[x - radius - 1];
                long bulk = acc * ww + (long) (lineIn[x - radius - 1] + lineIn[x + radius + 1]) * fw;
                lineOut[x] = (int) ((bulk + (1L << 23)) >> 24);
            }
            for (int x = edgeB; x <= lastx; x++) {
                acc += lineIn[lastx] - lineIn[x - radius - 1];
                long bulk = acc * ww + (long) (lineIn[x - radius - 1] + lineIn[lastx]) * fw;
                lineOut[x] = (int) ((bulk + (1L << 23)) >> 24);
            }
        } else {
            for (int x = 0; x < edgeB; x++) {
                acc += lineIn[x + radius] - lineIn[0];
                long bulk = acc * ww + (long) (lineIn[0] + lineIn[x + radius + 1]) * fw;
                lineOut[x] = (int) ((bulk + (1L << 23)) >> 24);
            }
            for (int x = edgeB; x < edgeA; x++) {
                acc += lineIn[lastx] - lineIn[0];
                long bulk = acc * ww + (long) (lineIn[0] + lineIn[lastx]) * fw;
                lineOut[x] = (int) ((bulk + (1L << 23)) >> 24);
            }
            for (int x = edgeA; x <= lastx; x++) {
                acc += lineIn[lastx] - lineIn[x - radius - 1];
                long bulk = acc * ww + (long) (lineIn[x - radius - 1] + lineIn[lastx]) * fw;
                lineOut[x] = (int) ((bulk + (1L << 23)) >> 24);
            }
        }
    }

    private static int[][] copyOf(int[][] src, int H, int W) {
        int[][] out = new int[H][W];
        for (int y = 0; y < H; y++) {
            System.arraycopy(src[y], 0, out[y], 0, W);
        }
        return out;
    }

    private static int[][] transpose(int[][] src, int H, int W) {
        int[][] out = new int[W][H];
        for (int y = 0; y < H; y++) {
            for (int x = 0; x < W; x++) {
                out[x][y] = src[y][x];
            }
        }
        return out;
    }

    private PilGaussianBlur() {}
}

File: hash/swift/RosettaSquintHash/Sources/RosettaSquintHash/Internal/PilGaussianBlur.swift
import Foundation

/// PIL-compatible GaussianBlur(radius=2) on UInt8 grayscale (row-major).
///
/// Pillow approximates a Gaussian via 3 separable box filters:
///   - n horizontal passes on the image
///   - n vertical passes (transpose → n horizontal passes → transpose back)
/// The box radius is derived from sigma via the Gwosdek et al. formula
/// (src/libImaging/BoxBlur.c in Pillow 10.4.0).
///
/// Boundary: edge replication (clamp to nearest valid pixel).
/// Intermediate and output pixels are rounded and clamped to [0, 255] as uint8
/// after every horizontal sweep pass.

func pilGaussianBlur(_ src: [UInt8], width: Int, height: Int, sigma: Double = 2.0, passes: Int = 3) -> [UInt8] {
    let floatRadius = gaussianBlurRadius(sigma: sigma, passes: passes)
    var current = src

    // n horizontal passes
    for _ in 0..<passes {
        var next = [UInt8](repeating: 0, count: width * height)
        for y in 0..<height {
            let rowIn  = Array(current[(y * width)..<(y * width + width)])
            let rowOut = horizontalBoxBlur(rowIn, floatRadius: floatRadius)
            for x in 0..<width {
                next[y * width + x] = rowOut[x]
            }
        }
        current = next
    }

    // n vertical passes: transpose → n horizontal passes → transpose back
    // Transposed: shape (width, height) stored row-major → transposed[x * height + y]
    var transposed = [UInt8](repeating: 0, count: width * height)
    for y in 0..<height {
        for x in 0..<width {
            transposed[x * height + y] = current[y * width + x]
        }
    }

    for _ in 0..<passes {
        var next = [UInt8](repeating: 0, count: width * height)
        // Each "row" in transposed has length = height (original height)
        for x in 0..<width {
            let rowIn  = Array(transposed[(x * height)..<(x * height + height)])
            let rowOut = horizontalBoxBlur(rowIn, floatRadius: floatRadius)
            for y in 0..<height {
                next[x * height + y] = rowOut[y]
            }
        }
        transposed = next
    }

    // Transpose back
    var result = [UInt8](repeating: 0, count: width * height)
    for y in 0..<height {
        for x in 0..<width {
            result[y * width + x] = transposed[x * height + y]
        }
    }
    return result
}

/// Converts GaussianBlur sigma + number of passes to the box filter fractional radius.
/// Implements `_gaussian_blur_radius` from Pillow's BoxBlur.c.
private func gaussianBlurRadius(sigma: Double, passes: Int) -> Double {
    let sigma2 = sigma * sigma / Double(passes)
    let L = (12.0 * sigma2 + 1.0).squareRoot()
    let l = (L - 1.0) / 2.0          // floor applied below
    let lFloor = floor(l)
    let aNum = (2.0 * lFloor + 1.0) * (lFloor * (lFloor + 1.0) - 3.0 * sigma2)
    let aDen = 6.0 * (sigma2 - (lFloor + 1.0) * (lFloor + 1.0))
    let a = aNum / aDen
    return lFloor + a
}

/// One horizontal box-blur pass over a single grayscale scanline.
/// Implements `ImagingHorizontalBoxBlur` (8-bit path) from Pillow's BoxBlur.c.
///
/// Fixed-point weights (matching the C source exactly):
///   ww = UInt32((1 << 24) / (floatRadius * 2 + 1))   // float division
///   fw = ((1 << 24) - (radius * 2 + 1) * ww) / 2      // integer division
///
/// Output per pixel: (bulk + (1 << 23)) >> 24  clamped to [0, 255].
private func horizontalBoxBlur(_ line: [UInt8], floatRadius: Double) -> [UInt8] {
    let n = line.count
    guard n > 0 else { return [] }
    let lastX = n - 1
    let radius = Int(floatRadius)  // floor / C cast-to-int

    // Fixed-point weights (UInt32 in C; use Int here to avoid overflow in acc * ww)
    let divisor = floatRadius * 2.0 + 1.0
    let ww = UInt32((Double(1 << 24)) / divisor)
    let innerPart = (1 << 24) - (radius * 2 + 1) * Int(ww)
    let fw = UInt32(innerPart / 2)

    let edgeA = min(radius + 1, n)
    let edgeB = max(n - radius - 1, 0)

    var lineOut = [UInt8](repeating: 0, count: n)

    // Initialize accumulator: pixels from -(radius+1) to +(radius-1) around x=-1.
    // Left clamped pixels (from -radius-1 to -1): all use line[0], count = radius+1
    // Right actual pixels (from 0 to edgeA-2): line[0..edgeA-2]
    // Remaining right clamped pixels (from edgeA-1 to radius-1 if edgeA <= radius):
    //   use line[lastX], count = radius - edgeA + 1
    var acc = UInt64(line[0]) * UInt64(radius + 1)
    for x in 0..<(edgeA - 1) {
        acc += UInt64(line[x])
    }
    acc += UInt64(line[lastX]) * UInt64(radius - edgeA + 1)

    func addFarSave(accVal: UInt64, left: Int, right: Int) -> UInt8 {
        let lc = max(0, min(lastX, left))
        let rc = max(0, min(lastX, right))
        let bulk = accVal &* UInt64(ww) &+ (UInt64(line[lc]) + UInt64(line[rc])) &* UInt64(fw)
        let shifted = Int64(bitPattern: (bulk &+ (1 << 23)) >> 24)
        if shifted < 0 { return 0 }
        if shifted > 255 { return 255 }
        return UInt8(shifted)
    }

    if edgeA <= edgeB {
        for x in 0..<edgeA {
            acc = acc &+ UInt64(line[min(x + radius, lastX)]) &- UInt64(line[0])
            lineOut[x] = addFarSave(accVal: acc, left: 0, right: x + radius + 1)
        }
        for x in edgeA..<edgeB {
            acc = acc &+ UInt64(line[x + radius]) &- UInt64(line[x - radius - 1])
            lineOut[x] = addFarSave(accVal: acc, left: x - radius - 1, right: x + radius + 1)
        }
        for x in edgeB..<n {
            acc = acc &+ UInt64(line[lastX]) &- UInt64(line[x - radius - 1])
            lineOut[x] = addFarSave(accVal: acc, left: x - radius - 1, right: lastX)
        }
    } else {
        for x in 0..<edgeB {
            acc = acc &+ UInt64(line[min(x + radius, lastX)]) &- UInt64(line[0])
            lineOut[x] = addFarSave(accVal: acc, left: 0, right: x + radius + 1)
        }
        for x in edgeB..<edgeA {
            acc = acc &+ UInt64(line[lastX]) &- UInt64(line[0])
            lineOut[x] = addFarSave(accVal: acc, left: 0, right: lastX)
        }
        for x in edgeA..<n {
            acc = acc &+ UInt64(line[lastX]) &- UInt64(line[x - radius - 1])
            lineOut[x] = addFarSave(accVal: acc, left: x - radius - 1, right: lastX)
        }
    }

    return lineOut
}
